### ISSUE TITLE
cmake: incorrect PKG_CONFIG_EXECUTABLE setting is used

### ISSUE THREAD

**Issue by uwehermann** [mxe/mxe#1023]
I was under the impression that an issue we're seeing was the same as #904, but I don't think that is the case anymore. Here's what seems to have been changed and goes wrong.

The setup:

A CMakeLists.txt file for a project is using something like this to find pkg-config-using libraries:
```
include(FindPkgConfig)
...
list(APPEND PKGDEPS libfoo>=x.y.z)
find_package(PkgConfig)
pkg_check_modules(PKGDEPS REQUIRED ${PKGDEPS})
```

Old (working) mxe.conf had this:
```set(PKG_CONFIG_EXECUTABLE /home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static-pkg-config)```

New (not working) mxe.conf has this:
```
file(GLOB mxe_cmake_files
    "/home/uwe/mxe-git/usr/i686-w64-mingw32.static/share/cmake/mxe-conf.d/*.cmake"
)
foreach(mxe_cmake_file ${mxe_cmake_files})
    include(${mxe_cmake_file})
endforeach()
```
And ```/home/uwe/mxe-git/usr/i686-w64-mingw32.static/share/cmake/mxe-conf.d/pkgconf.cmake``` contains:
```set(PKG_CONFIG_EXECUTABLE /home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static-pkg-config CACHE PATH "pkg-config executable")```

The problem/difference is the "CACHE" here, I think. With the above ```PKG_CONFIG_EXECUTABLE``` is ```/usr/bin/pkg-config``` before *and* after the ```set``` line. When dropping the ```CACHE PATH "pkg-config executable"``` part, the ```PKG_CONFIG_EXECUTABLE``` after the ```set``` line is now correct again:

```
/home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static-pkg-config
```

Using the correct ```PKG_CONFIG_EXECUTABLE``` the host cmake now finds the packages correctly again.

Please double-check if performing this change makes sense for MXE and all its requirements, I only checked one very specific use-case here.

**Comment by starius** [comment]
CACHE was added in 5a81e60e04efab090e4f90772666857260b4e2d3, #896

**Comment by tonytheodore** [comment]
From cmake docs:
>Since cache entries are meant to provide user-settable values this does not overwrite existing cache entries by default. Use the FORCE option to overwrite existing entries.

Is there somehow an existing cache entry for `PKG_CONFIG_EXECUTABLE`?

**Comment by uwehermann** [comment]
Ah, yes. I've looked a bit further and in this case the root cause is this snippet in our CMakeLists.txt file:

```
include(FindPkgConfig)
...
find_package(PkgConfig)
```

This leads to cmake checking for the pkg-config executable twice. Right now this results in:

```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.28") 
-- checking for module 'libsigrokcxx>=0.4.0'
--   package 'libsigrokcxx>=0.4.0' not found
```

With one of the fixes below the correct thing happens:
```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.28") 
-- Found PkgConfig: /.../i686-w64-mingw32.static.posix-pkg-config (found version "0.28") 
-- checking for module 'libsigrokcxx>=0.4.0'
--   found libsigrokcxx, version 0.4.0-git-8abdf00
```

Or we don't do the PkgConfig init twice in our software, then it works either way:
```
-- Found PkgConfig: /.../i686-w64-mingw32.static.posix-pkg-config (found version "0.28") 
-- checking for module 'libsigrokcxx>=0.4.0'
--   found libsigrokcxx, version 0.4.0-git-8abdf00
```

So here's the possible fixes:

Fix1:
```
-set(PKG_CONFIG_EXECUTABLE /.../i686-w64-mingw32.static.posix-pkg-config CACHE PATH "pkg-config executable")
+set(PKG_CONFIG_EXECUTABLE /.../i686-w64-mingw32.static.posix-pkg-config)
```

Fix2:
```
-set(PKG_CONFIG_EXECUTABLE /.../i686-w64-mingw32.static.posix-pkg-config CACHE PATH "pkg-config executable")
+set(PKG_CONFIG_EXECUTABLE /.../i686-w64-mingw32.static.posix-pkg-config CACHE PATH "pkg-config executable" FORCE)
```

Fix3:
Do nothing, declare multiple invokations of "include(FindPkgConfig)" and "find_package(PkgConfig)" in one CMakeLists.txt file a bug.

Either way, I think the intended behaviour in MXE should be defined and documented. Fix3 is something we should probably fix in our software either way, but OTOH it would also be nice if this worked regardless, i.e. Fix2 might be a nicer solution possibly. Assuming the FORCE there doesn't cause other problems, that is.

**Comment by tonytheodore** [comment]
Thanks for the investigation!

I'm thinking `FORCE` is what we want here, but I'm not clear on how the initial:
```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.28") 
```
happens if `cmake` is invoked with the toolchain file. Have you run `cmake` without the toolchain file previously? If so, then `FORCE` will likely break that workflow as you won't be able to switch back.

**Comment by uwehermann** [comment]
No, cmake is always run with the same toolchain file. Here's how to reproduce:

```
$ ls
CMakeLists.txt

$ cat CMakeLists.txt
cmake_minimum_required(VERSION 2.8.6)
include(FindPkgConfig)
project(foo)
list(APPEND PKGDEPS libsigrokcxx>=0.4.0)
find_package(PkgConfig)
pkg_check_modules(PKGDEPS REQUIRED ${PKGDEPS})

$ PKG_CONFIG_PATH_i686_w64_mingw32_static_posix=/home/uwe/sr_mingw/lib/pkgconfig:/home/uwe/mxe-git/usr/i686-w64-mingw32.static.posix/lib/pkgconfig cmake -DCMAKE_TOOLCHAIN_FILE=/home/uwe/mxe-git/usr/i686-w64-mingw32.static.posix/share/cmake/mxe-conf.cmake .
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.28") 
-- The C compiler identification is GNU 4.9.3
-- The CXX compiler identification is GNU 4.9.3
-- Check for working C compiler: /home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static.posix-gcc
-- Check for working C compiler: /home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static.posix-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static.posix-g++
-- Check for working CXX compiler: /home/uwe/mxe-git/usr/bin/i686-w64-mingw32.static.posix-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- checking for module 'libsigrokcxx>=0.4.0'
--   package 'libsigrokcxx>=0.4.0' not found
CMake Error at /usr/share/cmake-3.3/Modules/FindPkgConfig.cmake:340 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.3/Modules/FindPkgConfig.cmake:502 (_pkg_check_modules_internal)
  CMakeLists.txt:6 (pkg_check_modules)


-- Configuring incomplete, errors occurred!
```

The first "include(FindPkgConfig)" seems to do a pkg-config check (as does the second), but for some reason it finds another pkg-config and/or doesn't honor the toolchain file settings. I haven't really looked into that any further.

(Put any random *.pc file into ~/sr_mingw/lib/pkgconfig for testing and replace libsigrokcxx accordingly above)

**Comment by tonytheodore** [comment]
@ChristianFrisson, do you have any insights on this? Adding `FORCE` seems the way to go.

**Comment by ChristianFrisson** [comment]
Please remove `include(FindPkgConfig)`!

`find_package(PkgConfig)` is sufficient for the job.
cmake already knows where to find `FindPkgConfig.cmake` especially since it is a defaut search module, from the prefix where cmake is installed: `share/cmake-*/Modules/` (cmake version wildcarded here).

For more info check https://cmake.org/cmake/help/latest/command/find_package.html

I actually issued #896 otherwise multiple/first calls to `find_package(PkgConfig)` may fail.

Hoping this will help!

**Comment by tonytheodore** [comment]
> The first "include(FindPkgConfig)" seems to do a pkg-config check (as does the second), but for some reason it finds another pkg-config and/or doesn't honor the toolchain file settings.

Oddly, the behaviour seems to change if you put it after the `project(foo)` command. Setting the env var `PKG_CONFIG` works with later versions of cmake (but not 3.0.1 used by mxe) - I think `FORCE` is the way to go.
